### ISSUE TITLE
AOP-PHP can't recurse, but in big projects we need AOP advices to call other advices

### ISSUE THREAD

**Issue by tickleman** [AOP-PHP/AOP#36]
Please have a look to this example.
I want to force the day to begin before the man can wake up.
And I want to force the night to end before the day can begin.
All of this with AOP-PHP of course.

``` php
class Night {
    public static function end() { echo "the night ends "; }
}

class Day {
    public static function begin() { echo "the sun rises "; }
}

class Man {
    public function wakeup() { echo "the man wakes up "; }
}

aop_add_before("Day->begin()", array("Night", "end"));
aop_add_before("Man->wakeup()", array("Day", "begin"));

$me = new Man();
$me->wakeup();
```

But it won't work with current version 0.2.0 (my php is 5.4).
I am waiting for this result : <pre>the night ends the sun rises the man wakes up</pre>
But I only have this : <pre>the sun rises the man wakes up</pre>

Will AOP-PHP allow this ?

**Comment by tickleman** [comment]
notice : I added "static" to my sample, that had this little mistake sorry :).

(See you soon at the Afup's Php Tour at Nantes ! I promise I'll remove my pigtail before this date)

**Comment by geraldcroes** [comment]
We'll be pleased to see you there and to discuss around the extension !

**Comment by tickleman** [comment]
Additional test and example : only advices on functions or methods do not recurse. Advices on properties do recurse : 

``` php
class Night {   public $name = "NIGHT "; }

class Day {     public $name = "DAY "; }

class Man {     public $name = "MAN "; }

function getNightName() {   echo "tallest "; }

function getDayName() {     echo "up " . (new Night())->name; }

function getManName() {     echo "up " . (new Day())->name; }

aop_add_before("Night->name", "getNightName");
aop_add_before("Day->name", "getDayName");
aop_add_before("Man->name", "getManName");

echo (new Man())->name;
```

The output for this is correct : <pre>tallest up NIGHT up DAY MAN</pre>

**Comment by tickleman** [comment]
Example with simple fonctions :

``` php
function one() { echo "one "; }
function two() { echo "two "; }
function three() { echo "three "; }

function advice2() { one(); }
function advice3() { two(); }

aop_add_before("two()", "advice2");
aop_add_before("three()", "advice3");

three();
```

Should display : <pre>one two three</pre> Got : <pre>two three</pre>

---

Example with non-static object methods calls :

``` php
class One { public function aFunction() { echo "one "; } }

class Two { public function aFunction() { echo "two "; } }

class Three { public function aFunction() { echo "three"; } }

function advice2() { (new One())->aFunction(); }
function advice3() { (new Two())->aFunction(); }

aop_add_before("Two->aFunction()", "advice2");
aop_add_before("Three->aFunction()", "advice3");

(new Three())->aFunction();
```

Should display : <pre>one two three</pre> Got : <pre>two three</pre>

**Comment by fcrosfly** [comment]
I don't think this is a bug.

think about these code 

``` php
class A{

    function aa() {
        return 'aa';
    }

    function bb() {
        return 'bb' . $this->aa();
    }
}
$a = new A();
aop_add_around('A->aa()', array($a, 'bb'));
var_dump($a->aa());
```

**Comment by tickleman** [comment]
Yes, it will recurse. I can add controls to avoid recursion on this specific case.
But if I want to write highly modular application, I won't be abble to do this without AOP calls into AOP calls. I can't imagine different behaviours on a function if called inside or outside an advice.

``` php
class One { public function aFunction() { echo "one "; } }
class Two { public function aFunction() { echo "two "; } }
class Three { public function aFunction() { echo "three"; } }

function advice2() { (new One())->aFunction(); }
function advice3() { (new Two())->aFunction(); }

aop_add_before("Two->aFunction()", "advice2");
aop_add_before("Three->aFunction()", "advice3");

(new Two())->aFunction(); echo "<br>";
(new Three())->aFunction(); echo "<br>";
```

This case will display :

<pre>
one two
two three
</pre>

The Two::aFunction method will not do the same work the two times it is called ! I'm sure that does not comply with AOP standards.

And remember recursivity works with properties : that caused me some problems when trying to read a property value into the called advice on a "read" joinpoint on the same !... (look at https://github.com/AOP-PHP/AOP/issues/40). We must have the same behaviour on each AOP call, I think.

Last : I need it :). I can't imagine use AOP-PHP without this capability. I guess Julien is working, when I look at the last commits... I'm sure that https://github.com/AOP-PHP/AOP/commit/210d88d0962e1cfe728fd69c2d7ca6d6eb418f70 will help it.

**Comment by tickleman** [comment]
Let's try the "recursion" branch, which is supposed to solve this problem. My first test say "YESSSS IT DOES !". I will tell you more when more tests will have been done.

(notice : does not with "*()" wildcards yet).

**Comment by tickleman** [comment]
Are these cases of recursivity managed by the latest main branch build ?

I'm still working with the "recursion" branch you made for me (thanks a lot), and my project goes forward with it, but someday this will be great if I could test and work with the latest versions of AOP-PHP.

/yes i'm still alive/

**Comment by fcrosfly** [comment]
you can just comment aop_g(overxxx) = 1 to make it work.

2013/5/24 Tickleman notifications@github.com

**Comment by tickleman** [comment]
Well I commented aop_g-overxxx) = 1 but... it seems to crash, as a result.

I'm still working hard on my future framework, highly powered by my fork of an old version of AOP-PHP as a basis of its architecture... 'hope I will be abble to use the official release someday (x je croise les doigts x).

See you, very probably, at this year's end Afup's Forum PHP ;).

**Comment by baptistepillot** [comment]
I am very impatient to see this working in beta 3 !
